frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/contextmenu/PullSelectedBranchAction.java

```diff
@@ -0,0 +1,103 @@
+import com.virtuslab.gitmachete.frontend.actions.common.BasePullBranchAction;
+import com.virtuslab.gitmachete.frontend.datakeys.DataKeys;
+import com.virtuslab.logger.IPrefixedLambdaLogger;
+import com.virtuslab.logger.PrefixedLambdaLoggerFactory;
+
+/**
+ * Expects DataKeys:
+ * <ul>
+ *  <li>{@link DataKeys#KEY_GIT_MACHETE_REPOSITORY}</li>
+ *  <li>{@link DataKeys#KEY_SELECTED_BRANCH_NAME}</li>
+ *  <li>{@link CommonDataKeys#PROJECT}</li>
+ * </ul>
+ */
+public class PullSelectedBranchAction extends BasePullBranchAction {
+  private static final IPrefixedLambdaLogger LOG = PrefixedLambdaLoggerFactory.getLogger("frontendActions");
+
+  @Override
+  @UIEffect
+  public void update(AnActionEvent anActionEvent) {
+    super.update(anActionEvent);
+
+    Presentation presentation = anActionEvent.getPresentation();
+    if (!presentation.isEnabledAndVisible()) {
+      return;
+    }
+
+    Option<String> selectedBranchName = getSelectedBranchName(anActionEvent);
+
+    if (selectedBranchName.isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription("Pull disabled due to undefined current branch");
+      return;
+    }
+
+    Option<SyncToRemoteStatus> syncToRemoteStatus = getGitMacheteRepository(anActionEvent)
+        .flatMap(repo -> repo.getBranchByName(selectedBranchName.get()))
+        .map(branch -> branch.getSyncToRemoteStatus());
+
+    if (syncToRemoteStatus.isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription("Pull disabled due to undefined sync to remote status");
+      return;
+    }
+
+    SyncToRemoteStatus.Relation relation = syncToRemoteStatus.get().getRelation();
+    boolean isEnabled = PULL_ENABLING_STATUSES.contains(relation);
+
+    if (isEnabled) {
+      Option<Boolean> isSelectedEqualCurrent = getGitMacheteRepository(anActionEvent)
+          .flatMap(repo -> repo.getCurrentBranchIfManaged())
+          .map(branch -> branch.getName())
+          .map(branchName -> branchName.equals(selectedBranchName.get()));
+
+      if (isSelectedEqualCurrent.isDefined() && isSelectedEqualCurrent.get()) {
+        presentation.setText("Pull Current Branch");
+      }
+
+      presentation.setDescription("Pull branch '${selectedBranchName.get()}'");
+
+    } else {
+      presentation.setEnabled(false);
+      String description = getRelationBaseDescription(relation);
+      presentation.setDescription(description);
+    }
+
+  }
+
+  @Override
+  @UIEffect
+  public void actionPerformed(AnActionEvent anActionEvent) {
+    LOG.debug("Performing");
+
+    Project project = getProject(anActionEvent);
+    Option<GitRepository> selectedVcsRepository = getSelectedVcsRepository(anActionEvent);
+    Option<String> branchName = getSelectedBranchName(anActionEvent);
+
+    if (branchName.isDefined()) {
+      if (selectedVcsRepository.isDefined()) {
+        doPull(project, selectedVcsRepository.get(), branchName.get());
```

Comment:
Actually... what happens for the following sequence of events:
1) local: `refs/heads/foo` = X~2, `refs/remotes/origin/foo` = X => sync to remote sync status == `BehindRemote` => pull allowed
2) remote: in the meantime, `foo` is force-pushed to point to commit Y, which is **not** descendant of X~2
3) local: fetching Y as `refs/remotes/origin/foo`; fast-forwarding `refs/heads/foo` from X~2 to Y not possible

Reply:
Ideally, it should behave in `git pull --ff-only`-like manner, so just fetch `refs/remotes/origin/foo` but reject updating `refs/heads/foo` if FF isn't possible (+ warn/error message)

Reply:
So this is what happens:
`15:58:22.909: [filter-service] git -c credential.helper= -c core.quotepath=false -c log.showSignature=false fetch origin develop:develop --recurse-submodules=no --progress --prune`
![image](https://user-images.githubusercontent.com/19799111/81829404-2388d300-953b-11ea-8e4a-18f0cfc63701.png)

Reply:
So the `refspec` is now:
```
 var localFullName = trackingInfo.getLocalBranch().getFullName();
 var remoteFullName = trackingInfo.getRemoteBranch().getFullName();
 var refspec = "+${localFullName}:${remoteFullName}";
```